File: Cajero.py
#Ejemplo de clases
#class figura:#se define la clase de figura
    
    #def __init__(self,num_lados,longitud):
        #self.num_lados=num_lados
        #self.longitud=longitud

    #def hallar_perimetro(self):
        #return self.num_lados*self.longitud
#piramide=figura(4,8)
#res=piramide.hallar_perimetro()

#Ejemplo usuario con clases

class Usuario:

    def __init__(self, nombre, pin, saldo):
        self.nombre=nombre
        self.pin=pin
        self.saldo=saldo

class Banco:

    def __init__(self, usuarios=[]):
        self.usuarios=usuarios
    def autenticar(self, nombre, pin):
        for usuario in self.usuarios:
            if usuario.nombre==nombre:
                if usuario.pin==pin:
                    print('Estas Logeado')
                    return True
                else:
                    print('Pin o nombre incorrecto')
                    return False
        else:
            print('El usuario no existe')
            return False
    def sacar_dinero(self, nombre, saldo):
        for usuario in self.usuarios:
            if usuario.nombre==nombre:
                if usuario.saldo<saldo:
                    print('Saldo insuficiente')
                    break
                elif usuario.saldo>=saldo:
                    usuario.saldo-=saldo
                    print('El saldo disponible es ',usuario.saldo)

ana = Usuario('Ana',9872,450)
pablo = Usuario('Pablo',5555,200)
rodrigo=Usuario('Rodrigo',2222,900)

banco = Banco(usuarios=[ana,pablo,rodrigo])
while True:
    print('Bienvenido al Banco, por favor, identifiquese.')
    print('Introduzca el nombre.')
    nombre =input()
    print('Introduzca el pin')
    pin =int(input())

    if banco.autenticar(nombre,pin):
        while True:
            print('Por favor, elija una de estas opciones:\n 1. Sacar dinero\n 2.Terminar sesion.')
            opcion=input()

            if opcion =='1':
                print('Introduce cantidad a sacar.')
                saldo=int(input())
                banco.sacar_dinero(nombre, saldo)
            
            elif opcion=='2':
                print('Saliendo del sistema.')
                break
            else:
                print('Opcion incorrecta. Por favor, introduzca otra opcion.')
    else:
        print('Usuario no autenticado. Por favor, Introduzca nombre y pin correctos.')

File: Cajero.js
const readline = require("readline"); // Importamos el módulo readline para manejar la entrada del usuario en la consola

// Clase Usuario que representa a un cliente del banco
class Usuario {
    constructor(nombre, pin, saldo) {
        this.nombre = nombre; // Nombre del usuario
        this.pin = pin; // PIN de acceso
        this.saldo = saldo; // Saldo en la cuenta
    }
}

// Clase Banco que maneja los usuarios y sus operaciones
class Banco {
    constructor(usuarios = []) {
        this.usuarios = usuarios; // Lista de usuarios registrados en el banco
    }

    // Método para autenticar al usuario
    autenticar(nombre, pin) {
        for (let usuario of this.usuarios) {
            if (usuario.nombre === nombre) { // Comprobamos si el usuario existe
                if (usuario.pin === pin) { // Verificamos si el PIN es correcto
                    console.log("¡Estás logueado!");
                    return true;
                } else {
                    console.log("Pin o nombre incorrecto.");
                    return false;
                }
            }
        }
        console.log("El usuario no existe.");
        return false;
    }

    // Método para retirar dinero de la cuenta de un usuario
    sacarDinero(nombre, cantidad) {
        for (let usuario of this.usuarios) {
            if (usuario.nombre === nombre) {
                if (usuario.saldo < cantidad) { // Verificamos si el saldo es suficiente
                    console.log("Saldo insuficiente.");
                } else {
                    usuario.saldo -= cantidad; // Restamos la cantidad al saldo
                    console.log("El saldo disponible es: " + usuario.saldo);
                }
                return;
            }
        }
    }
}

// Crear instancias de usuarios con nombre, PIN y saldo inicial
const ana = new Usuario("Ana", 9872, 450);
const pablo = new Usuario("Pablo", 5555, 200);
const rodrigo = new Usuario("Rodrigo", 2222, 900);

// Crear una instancia del Banco con los usuarios registrados
const banco = new Banco([ana, pablo, rodrigo]);

// Configuración del readline para manejar la entrada y salida de datos en consola
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Función para iniciar sesión en el sistema
function iniciarSesion() {
    rl.question("Introduzca su nombre: ", (nombre) => {
        rl.question("Introduzca su pin: ", (pin) => {
            pin = parseInt(pin); // Convertimos el PIN a número
            if (banco.autenticar(nombre, pin)) { // Si la autenticación es exitosa, mostramos el menú
                menuUsuario(nombre);
            } else {
                iniciarSesion(); // Si falla, pedimos los datos nuevamente
            }
        });
    });
}

// Función que muestra el menú de opciones para el usuario autenticado
function menuUsuario(nombre) {
    rl.question("Por favor, elija una opción:\n1. Sacar dinero\n2. Terminar sesión\n", (opcion) => {
        if (opcion === "1") { // Opción para retirar dinero
            rl.question("Introduce cantidad a sacar: ", (cantidad) => {
                banco.sacarDinero(nombre, parseInt(cantidad)); // Llamamos al método para retirar dinero
                menuUsuario(nombre); // Volvemos a mostrar el menú
            });
        } else if (opcion === "2") { // Opción para cerrar sesión
            console.log("Saliendo del sistema.");
            rl.close(); // Cerramos la interfaz readline
        } else { // Opción inválida
            console.log("Opción incorrecta.");
            menuUsuario(nombre); // Volvemos a mostrar el menú
        }
    });
}

// Mensaje de bienvenida e inicio de sesión
console.log("Bienvenido al Banco");
iniciarSesion();
